tracer/src/Datadog.Trace/Logging/DirectSubmission/Formatting/LogFormatter.cs

```diff
@@ -0,0 +1,233 @@
+﻿// <copyright file="LogFormatter.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Globalization;
+using System.IO;
+using System.Text;
+using Datadog.Trace.Vendors.Newtonsoft.Json;
+
+namespace Datadog.Trace.Logging.DirectSubmission.Formatting
+{
+    internal class LogFormatter
+    {
+        private const string SourcePropertyName = "ddsource";
+        private const string ServicePropertyName = "ddservice";
+        private const string HostPropertyName = "host";
+        private const string TagsPropertyName = "ddtags";
+        private const string EnvPropertyName = "dd_env";
+        private const string VersionPropertyName = "dd_version";
+
+        private readonly string _source;
+        private readonly string _service;
+        private readonly string _host;
+        private readonly string _globalTags;
+        private readonly string _env;
+        private readonly string _version;
+
+        public LogFormatter(
+            DirectLogSubmissionSettings settings,
+            string serviceName,
+            string env,
+            string version)
+        {
+            _source = string.IsNullOrEmpty(settings.Source) ? null : settings.Source;
+            _service = string.IsNullOrEmpty(serviceName) ? null : serviceName;
+            _host = string.IsNullOrEmpty(settings.Host) ? null : settings.Host;
+            _globalTags = string.IsNullOrEmpty(settings.GlobalTags) ? null : settings.GlobalTags;
+            _env = string.IsNullOrEmpty(env) ? null : env;
+            _version = string.IsNullOrEmpty(version) ? null : version;
+        }
+
+        internal static bool IsSourceProperty(string propertyName) =>
+            string.Equals(propertyName, SourcePropertyName, StringComparison.OrdinalIgnoreCase);
+
+        internal static bool IsServiceProperty(string propertyName) =>
+            string.Equals(propertyName, ServicePropertyName, StringComparison.OrdinalIgnoreCase)
+         || string.Equals(propertyName, "dd_service", StringComparison.OrdinalIgnoreCase)
```

Comment:
why this special case for service and not the others?

Reply:
There are 3 ways to add the `Service` tag (`ddservice`, `dd_service`, `dd.service`) depending on when it's added. There are 2 ways for Environment (`dd_env`, `dd.env`) and 2 for Version (`dd.version`, `dd_version`) and only one for the others, hence all this annoying faff 😉